Automattic/simplenote-macos
# Implementing NotesListController: Mark III

### Details
In this PR we're dropping NSArrayController, in favor of **NotesListController / ResultsController** 🔥 

@eshurakov (Relatively small PR), but complex to the bones.
I've tried to further split this one, but no matter what, it's a huge change. (Suggestions always welcome!!).

Thanks in advance!!!

Closes #102
Ref. #626


---

### Test: Launching
- [ ] Verify the First Note is preselected
- [ ] Verify the Editor properly displays the first note
- [ ] Verify the Editor's Top Insets look great

### Test: Interface
- [ ] Verify the `New Note` Toolbar Action is disabled when you're visualizing the Trash
- [ ] Verify the `New Note` Toolbar Action is enabled when you're **NOT** visualizing the Trash
- [ ] Verify Switching to / from Light / Dark Mode looks super, and your Note List selection is preserved
- [ ] Verify clicking over an empty tag causes the `No Notes` legend to show up in the Notes List
- [ ] Verify the Notes List Title reflects the currently selected Tag
- [ ] Verify CMD + Clicking over multiple notes causes the Editor to display the `XX Notes Selected` legend
- [ ] Verify Right Clicking over any Notes Row displays a Menu that allows you to Trash / Pin / Unpin / Copy Interlink
- [ ] Verify Right Clicking over any **Trashed** Note Row displays a different menu, that allows you to Delete Forever / Restore
- [ ] Verify the Notes List gets immediately reordered whenever you're sorting by Note Contents, and the first character is modified

### Test: Search
1. Launch the app and click over the Search Button

- [ ] Verify entering any keyword filters the Notes List
- [ ] Verify clearing the Search Field ""Unfilters" the Notes List
- [ ] Verify clicking over the (x) button dismisses the search, and "Unfilters" the Notes List

### Test: Trash
1. Launch the app
2. Add +5 Notes and Trash Them
3. Click over the Trash row (Tags List), and select the Last Row

- [ ] Verify that deleting the last visible row (at the bottom) causes no exception in the console
- [ ] Verify that immediately after deleting the last row, `N - 1` is preselected
- [ ] Verify that after deleting the row at the top, whatever row ends up on top is preselected
- [ ] Verify that Emptying the Trash causes the Editor's contents to get cleared out

### Test: Reordering Tags
1. Launch the app and select a random Tag with +3 Notes
2. Click over any note that shows up in the middle of the screen

- [ ] Verify that switching from Light <> Dark Mode causes the app to refresh, **but the app preserves the selected Note**
- [ ] Click over `View > Tag Sorting > Alphabetically` and verify the Notes List gets updated, matching the selected Tag

### Test: Updating
1. Launch the app and select a random Tag with +3 Notes
2. Click over any note that shows up in the middle of the screen
3. In another device, open the same note, and remove such Tag

- [ ] Verify the UI is immediately reloaded
- [ ] Verify a new Note (at the same index) is preselected 

### Test: All Notes Selection
1. Add a New Note into an (Empty) New `Tag A`
2. Add a reference to such Note from any other Note (that's not in `Tag A`)
3. Click over `Tag A`
4. Click over the top right `( i )` icon and click over any Inbound Reference

- [ ] Verify the Tags List jumps immediately to`All Notes` 
- [ ] Verify the Notes List displays all of your notes
- [ ] Verify the Editor now renders the referencing document

### Test: Deletion
1. Launch the App
2. Press CMD + Click over multiple notes and press **backspace**

- [ ] Verify the highlighted notes in (2) are trashed

### Test: Realtime Sync
1. Launch the App and click over Trash
2. Click `Empty Trash` and make sure the list is clear
3. In another device, create a new Note and trash it

- [ ] Verify a new Note shows up, with an animation, and gets preselected

### Release
`RELEASE-NOTES.txt` was updated in dcc9e0d with:

> Notes List will now animate content changes #764
